Restructure HomeScreen

devendrn opened the pull request:
It doesn't make much sense to show notification banners in Unlock, Changelog, and Unsupported screens, so I'm moving them (Gamepad, Update check banners) to Home screen. Same thing for download progress bar, only useful in home screen.

I have additional changes to add. Might use this PR to request review in certain parts of the code.

devendrn committed:
Move all notification banners to HomeScreen

devendrn committed:
Simplify some logic

devendrn commented:
Replaced some for loops with more readable alternatives. <https://doc.qt.io/qt-6/qtqml-javascript-functionlist.html#array-objects-array-prototype-object-function-properties-15>

ChristopherHX commented on qml/HomeScreen.qml:
```diff
@@ -234,14 +259,51 @@ BaseScreen {
             property bool canDownload: googleLoginHelper.account !== null && playVerChannel.licenseStatus == 3
+            property bool statusChecking: !isVersionsInitialized || playVerChannel.licenseStatus === 0 || playVerChannel.licenseStatus === 1
+
+            function getPlayLabel() {
+                if (statusChecking)
+                    return qsTr("Please wait...")
+
+                if (!((playVerChannel.hasVerifiedLicense || launcherSettings.trialMode || !LAUNCHER_ENABLE_GOOGLE_PLAY_LICENCE_CHECK) && (canDownload || !needsDownload())))
+                    return googleLoginHelper.account !== null ? qsTr("Ask Google Again") : qsTr("Sign In")
+
+                if (gameLauncher.running)
+                    return qsTr("Game is running")
+
+                if (!checkSupport())
+                    return qsTr("Unsupported Version")
+
+                if (needsDownload()) {
+                    if (googleLoginHelper.account === null) {
+                        return qsTr("Sign in")
+                    }
+
+                    if (profileManager.activeProfile.versionType === ProfileInfo.LATEST_GOOGLE_PLAY && googleLoginHelper.hideLatest)
+                        return qsTr("Please sign in again")
+
+                    return qsTr("Download and play")
+                }
+
+                return qsTr("Play")
+            }
```
yes this looks like a good idea to move it to a function.

I guess that I didn't know in 2019/20 that qt supports binding updates written in functions.

ChristopherHX replied:
Maybe we should go one step further and merge text, subtext and enabled into a single function returning an js object.

Store the result in a custom property field then assign the three fields via this object.

I did duplicate way too much logic of the play button.

Now the question is does qt qml support my idea or would it fail at runtime?

What do you think?

devendrn replied:
I'll try that and let you know

devendrn replied:
Using a js object to assign multiple fields together seems to complicate code even further. Keeping everything separate for the current logic looks easier to read compared to merged one.

Heres how it looks when I tried to merge just text and subtext logic.
```qml
            property var label: {
                var label = {
                    "text": "",
                    "subtext": ""
                }

                if (statusChecking) {
                    return label
                }

                const displayedVersionName = getDisplayedVersionName()
                label.subtext = displayedVersionName ? ("Minecraft " + displayedVersionName).toUpperCase() : qsTr("Please wait")

                if (gameLauncher.running) {
                    label.text = qsTr("Game is running")
                    return label
                }

                if (googleLoginHelper.account === null) {
                    label.text = qsTr("Sign in")
                    label.subtext = ""
                    return label
                }

                if (!playVerChannel.hasVerifiedLicense && LAUNCHER_ENABLE_GOOGLE_PLAY_LICENCE_CHECK && !launcherSettings.trialMode) {
                    label.text = qsTr("Ask Google Again")
                    label.subtext = qsTr("Failed to obtain apk url")
                }

                if (!displayedVersionName) {
                    return label
                }

                if (!checkSupport()) {
                    label.text = qsTr("Unsupported Version")
                    return label
                }

                if (needsDownload()) {
                    if (profileManager.activeProfile.versionType === ProfileInfo.LATEST_GOOGLE_PLAY && googleLoginHelper.hideLatest) {
                        label.text = qsTr("Please sign in again")
                        label.subtext = ""
                        return label
                    }
                    label.text = qsTr("Download and play")
                    return label
                }

                label.text = qsTr("Play")
                return label
            }

            text: label.text.toUpperCase()
            subText: label.subtext
```

Separated logic:
```qml
            text: {
                if (statusChecking)
                    return ""

                if (gameLauncher.running)
                    return qsTr("Game is running").toUpperCase()

                if (googleLoginHelper.account === null)
                    return qsTr("Sign in").toUpperCase()

                if (!playVerChannel.hasVerifiedLicense && LAUNCHER_ENABLE_GOOGLE_PLAY_LICENCE_CHECK && !launcherSettings.trialMode)
                    return qsTr("Ask Google Again").toUpperCase()

                if (!displayedVersionName)
                    return ""

                if (!checkSupport())
                    return qsTr("Unsupported Version").toUpperCase()

                if (needsDownload()) {
                    if (profileManager.activeProfile.versionType === ProfileInfo.LATEST_GOOGLE_PLAY && googleLoginHelper.hideLatest)
                        return qsTr("Please sign in again").toUpperCase()
                    return qsTr("Download and play").toUpperCase()
                }

                return qsTr("Play").toUpperCase()
            }

            subText: {
                if (statusChecking)
                    return ""

                if (displayedVersionName)
                    return ("Minecraft " + displayedVersionName).toUpperCase()

                if (googleLoginHelper.account === null || (needsDownload() && profileManager.activeProfile.versionType === ProfileInfo.LATEST_GOOGLE_PLAY && googleLoginHelper.hideLatest))
                    return ""

                return qsTr("Please wait")
            }

            enabled: !(gameLauncher.running || statusChecking || playDownloadTask.active || apkExtractionTask.active || updateChecker.active || !checkSupport() || !getDisplayedVersionName())
```

ChristopherHX replied:
Oh yes you are right, feel free to use Separated logic.

ChristopherHX reviewed (commented):
I think that preferring loops over includes is due to me writing code in so many programming languages that share the loop construct, but not the includes method.

All fine to use includes, just saying it's unlikely for me to use it intentionally.

_Need to test this on my old laptop this weekend, never set up the UI dev environment on my new device._

devendrn committed:
Update playbutton

- Add waiting animation
- Update button logic

devendrn committed:
Update notification banners

devendrn committed:
Fix focus border issues with notification buttons

devendrn committed:
Update default background art

devendrn committed:
Separate play button logic

- Move actions other than download/play into notify banner
- Other minor changes

devendrn commented:
@ChristopherHX I've made some assumptions regarding some of the logic. Please clarify if they are incorrect.
- Must be signed into a Google Play account to use the launcher (for development release and final release).
- Must have a valid license to download/play game. Can be bypassed by a compile flag, but only allows downloading when license is valid.

The following is a summary of all changes made in this PR.

### BaseScreen
- Progress bar and Banners were moved to HomeScreen. UnlockScreen,  UnsupportedScreen, and ChangelogScreen don't need them. 

### NotifyBanner Component
- Banners are now a padded box aligned to top-center of background art.
- Banners can be dismissed (Not persistent, as there isn't much advantage in making it so).

### PlayButton
- Use loading bar animation instead of showing wait text.
- Only handle play and download cases. Moved other logic into banners section.

### Banners
- Banner area is scrollable if content area exceeds available space.
- Added some extra actions:
  - Go to game log screen.
  - Open gamepadTool window.
- **TODO (Feedback needed)**: Do not hide main playStatusNotify banner (which now handles logic moved from the playButton). Or remove the option to hide banners as a whole from settings?

### Home screen
- **(Feedback needed)** New default background (takes extra 22.1kb bin space in repo though).
- Disabled profile edit components when not signed in.
- Disabled smoothing for small background images (pixel arts).

### Other changes
- Rewrote some texts.
- Renamed some names for consistency.
- Used built-in qml js functions to simplfy some loop code.
- Made some minor UI adjustments.

ChristopherHX commented:
> * Must be signed into a Google Play account to use the launcher (for development release and final release).

Originally bypass included the login requirement, but I don't care about the bypass functionality. It is just for the case google login becomes impossible and is irreparable.
We could remove the cmake + main.cpp code to hide it and just keep qml support, but again I don't care.

> (Feedback needed) New default background (takes extra 22.1kb bin space in repo though).

Looks good, it's ok for me to embed it here, if we don't change it frequently.


> * **TODO (Feedback needed)**: Do not hide main playStatusNotify banner (which now handles logic moved from the playButton). Or remove the option to hide banners as a whole from settings?

I need to look into your PR this weekend I don't know right now, too busy right now.

devendrn committed:
Undo a bad change

ChristopherHX commented:
Sorry for the delay, I hope to allocate a bit time for review tomorrow. Had other projects on my radar...

At this point I have finally set up a qt6 environment on my new pc to compile this project.

_Offtopic_ I wonder did GitHub close this PR by your force push automatically? Had this feeling once somewhere else.

devendrn commented:
No worries. I have no intention of wanting you to review this soon. 

> Offtopic I wonder did GitHub close this PR by your force push automatically? Had this feeling once somewhere else.

During a part of my rebase, I accidentally pushed a branch matching origin/qt6 head. PR closed automatically since there was no change.

ChristopherHX commented:
I have Minecraft Beta installed via dev-mode, the built of this draft shows incorrectly `Download and Play` and behaves like the message even if the game has been installed and is runnable for me without downloading anything (using an older version of the UI).


> Do not hide main playStatusNotify banner

This is a good idea to never hide them, as this one has severity critical.

Maybe just make this dependent to `dismissable=false`, I fully agree with this.

I still assume hiding dismissible notifications via settings is a good thing.

Spellchecker, are telling me `dismissable` is wrong and should be written like `dismissible`. (my english skills wouldn't have this flagged)

devendrn commented:
I'll look into it and update you.

devendrn committed:
Fix typo

devendrn committed:
Hide only less critical notify banners

devendrn committed:
Fix typo

devendrn commented:
> I have Minecraft Beta installed via dev-mode, the built of this draft shows incorrectly Download and Play...

Typo from my point. Should be fixed now. Please check again.

> Maybe just make this dependent to dismissable=false, I fully agree with this.

I have added `&& launcherSettings.showNotifications` to all notify banners beside `playStatusNotify`. Alternatively I could put them all in a `Column` and hide that whole together, but that would add an extra indentation.

ChristopherHX commented:
> Typo from my point. Should be fixed now. Please check again.

Yes seems to work now 👍 .

_Given how much error paths exist in my original code without automated testing this is going to be an adventure_

Scenario
- Google Login loaded from previous version: ok
- Run existing Download: ok
- Find the gamelog after letting the game crash due to missing libEGL.dylib: Not ok, but why? (it worked for sure in the qt5 version)
